### DISCUSSION TITLE
Compiling an external CMake project

### DISCUSSION THREAD

**Discussion by bettio** [zephyrproject-rtos/zephyr#65704] [General]
We are trying to build a cmake project that is required from our application, but we either need to implement it in a workaround-ish way or we get an error.

So I'm looking for suggestions about how to correctly implement this.

### Approach 1 (which causes an error):
```cmake
cmake_minimum_required(VERSION 3.20.0)

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(zephyr_avm_platform_app)

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../../../CMakeModules")
add_subdirectory("${CMAKE_CURRENT_SOURCE_DIR}/../../libAtomVM" "build")

target_sources(app PRIVATE src/main.c)
```

causes the following error:
```
CMake Error at /home/davide/Projects/AtomVM/src/libAtomVM/CMakeLists.txt:100 (target_compile_features):
  target_compile_features The compiler feature "c_std_11" is not known to C
  compiler

  "GNU"

  version 12.2.0.
  ```

### Approach 2 (which looks like a workaround):

Basically the content of `${CMAKE_CURRENT_SOURCE_DIR}/../../libAtomVM/CMakeLists.txt` is duplicated on a "zephyr specialized" cmake file e.g. `src/platforms/zephyr/src/libAtomVM/CMakeLists.txt`.
I feel this is quite a work around since every change in the original cmake file has to be copied&pasted into the new file.

**Comment by nordicjm** [comment]
Have you looked at this https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/application_development/external_lib ?

**Reply by bettio** [reply]
I looked at it, but it felt like the wrong solution, since cmake documentation says:

> The ExternalProject_Add() function creates a custom target to drive download, update/patch, configure, build, install and test steps of an external project:

So my understanding is that `ExternalProject_Add()` works for using a library from a git repository somewhere in another place.

What I'm trying to do is building a common `.a` library that is located in another directory of the same source tree which has its own [CMakeLists.txt](https://github.com/atomvm/AtomVM/tree/master/src/libAtomVM).

Basically we are trying adding to the project the Zephyr platform support, that makes use of existing code, without changing  or duplicating stuff all around in order to support it.

We used doing this with ESP-IDF port with `add_subdirectory("${CMAKE_CURRENT_SOURCE_DIR}/../../../../libAtomVM" "libAtomVM")`, that was compiling that common `.a` without any additional issue.

**Reply by nordicjm** [reply]
If you mean just including code, then use `add_subdirectory()` or make it a west module and add the files to the build like that. If you want it compiled separately as a library, then included as a static library into the zephyr build as a second build, you need to use the method shown in the example above.

**Reply by bettio** [reply]
`add_subdirectory()` is the approach I described in "Approach 1" and it causes a `target_compile_features The compiler feature "c_std_11" is not known to C compiler` error.
